kernel/src/mem.rs

```diff
@@ -1,202 +1,278 @@
-    /// beginning of the AppSlice.
-    ///
-    /// See
-    /// [`std::slice::chunks_mut()`](https://doc.rust-lang.org/std/primitive.slice.html#method.chunks_mut).
-    ///
-    /// Internally this uses
-    /// [`AsMut`](struct.AppSlice.html#impl-AsMut<[T]>), hence when
-    /// the app dies, restarts or the
-    /// [`AppId`](crate::callback::AppId) changes for any other
-    /// reason, a [`ChunksMut`](core::slice::ChunksMut) iterator of zero length will
-    /// be returned.
-    ///
-    /// # Panics
-    ///
-    /// Panics if `chunk_size` is 0.
-    pub fn chunks_mut(&mut self, size: usize) -> slice::ChunksMut<T> {
-        self.as_mut().chunks_mut(size)
+    pub unsafe fn new_external(
+        ptr: *const u8,
+        len: usize,
+        process_id: AppId,
+        _cap: &dyn capabilities::ExternalProcessCapability,
+    ) -> Self {
+        Self::new(ptr, len, process_id)
+    }
+
+    pub(crate) fn consume(self) -> (*const u8, usize) {
+        (self.ptr, self.len)
     }
 }
 
-impl<L, T> AsRef<[T]> for AppSlice<L, T> {
-    /// Get a slice reference over the userspace buffer
-    ///
-    /// This first checks whether the app died, restarted, or its
-    /// AppId identifier changed for any other reason. In this case, a
-    /// slice of length zero is returned.
-    fn as_ref(&self) -> &[T] {
-        self.ptr
-            .process
-            .kernel
-            .process_map_or(&[], self.ptr.process, |_| unsafe {
-                slice::from_raw_parts(self.ptr.ptr.as_ref(), self.len)
-            })
+impl Default for ReadOnlyAppSlice {
+    fn default() -> Self {
+        ReadOnlyAppSlice {
+            ptr: 0x0 as *mut u8,
+            len: 0,
+            process_id: None,
+        }
     }
 }
 
-impl<L, T> AsMut<[T]> for AppSlice<L, T> {
-    /// Get a mutable slice reference over the userspace buffer
-    ///
-    /// This first checks whether the app died, restarted, or its
-    /// AppId identifier changed for any other reason. In this case, a
-    /// slice of length zero is returned.
-    fn as_mut(&mut self) -> &mut [T] {
-        self.ptr
-            .process
-            .kernel
-            .process_map_or(&mut [], self.ptr.process, |_| unsafe {
-                slice::from_raw_parts_mut(self.ptr.ptr.as_mut(), self.len)
-            })
+impl Read for ReadOnlyAppSlice {
+    fn len(&self) -> usize {
+        self.process_id
+            .map_or(0, |pid| pid.kernel.process_map_or(0, pid, |_| self.len))
+    }
+
+    fn ptr(&self) -> *const u8 {
+        if self.len == 0 {
+            0x0 as *const u8
+        } else {
+            self.ptr
+        }
+    }
+
+    fn map_or<F, R>(&self, default: R, fun: F) -> R
+    where
+        F: FnOnce(&[u8]) -> R,
+    {
+        match self.process_id {
+            None => default,
+            Some(pid) => pid.kernel.process_map_or(default, pid, |_| {
+                let slice = unsafe { slice::from_raw_parts(self.ptr, self.len) };
```

Comment:
Needs `unsafe` documentation (maybe copy of rational from `mut_map_or` above?)

Reply:
This is in #2464.